src/babel-serializer/network/pcap.hh

```diff
@@ -4,13 +4,142 @@
+    /// allocating constructor that will read the whole file into memory
+    reader(cc::string_view filepath, error_handler on_error = default_error_handler)
+    {
+        _on_error = on_error;
+        _file_data = file::read_all_bytes(filepath, on_error); // should this be lazy instead?
+        _data = _file_data.data();
+        _size = _file_data.size();
+    }
+    /// non-allocating constructor that references into some other data
+    reader(cc::span<std::byte const> data, error_handler on_error = default_error_handler)
+    {
+        _on_error = on_error;
+        _data = data.data();
+        _size = data.size();
+    }
+
+private:
+    struct packet_iterator
+    {
+        struct packet_header
+        {
+            uint32_t ts_sec;   /* timestamp seconds */
+            uint32_t ts_usec;  /* timestamp microseconds */
+            uint32_t incl_len; /* number of octets of packet saved in file */
+            uint32_t orig_len; /* actual length of packet */
+        };
+
+        packet_iterator(cc::span<std::byte const> data, error_handler on_error) : _data{data}, _on_error{on_error}
+        {
+            if (!_data.empty())
+                _current_packet = read_packet();
+        }
+
+        packet operator*() { return _current_packet; }
+
+        packet_iterator& operator++()
+        {
+            _pos += sizeof(packet_header) + _current_packet.data.size();
+            if (_pos < _data.size())
+                _current_packet = read_packet();
+            return *this;
+        }
+
+        packet_iterator operator++(int)
+        {
+            auto i = *this;
+            operator++();
+            return i;
+        }
+
+        bool operator!=(cc::sentinel) const { return _pos < _data.size(); }
+
+    private:
+        packet read_packet();
+        cc::span<std::byte const> _data;
+        packet _current_packet;
+        size_t _pos = 0;
+        error_handler _on_error;
+    };
+
+    struct packet_range
+    {
+        packet_range(cc::span<std::byte const> data, error_handler on_error) : data{data}, on_error{on_error} {}
+
+        cc::span<std::byte const> data; ///< data without the header
+        error_handler on_error;
+        packet_iterator begin() const { return {data, on_error}; }
+        cc::sentinel end() const { return {}; }
+    };
+
+public: // data
+    header header()
+    {
+        struct header h;
+        if (_size < sizeof(h))
+        {
+            _on_error(cc::span(_data, _size), cc::span(_data, _size).last(0), "Unexpected end of file", severity::error);
+            return {};
+        }
+        cc::as_byte_span(h).copy_from(cc::as_byte_span(_data, 0, sizeof(h)));
+        return h;
+    }
+
+    packet_range packets() { return {cc::span(_data, _size).subspan(sizeof(struct header)), _on_error}; }
+
+    std::byte const* data() const { return _data; }
+    size_t size() const { return _size; }
+
+private:
+    std::byte const* _data = nullptr;
+    size_t _size = 0;
```

Comment:
This could be a span<byte>, making `_size` less ambiguous (ie., is it the amount of packets?)